workflows/seqinspector.nf

```diff
@@ -50,45 +57,162 @@ workflow SEQINSPECTOR {
+    summary_params                        = paramsSummaryMap(
         workflow, parameters_schema: "nextflow_schema.json")
-    ch_workflow_summary = Channel.value(paramsSummaryMultiqc(summary_params))
-
+    ch_workflow_summary                   = Channel.value(
+        paramsSummaryMultiqc(summary_params))
     ch_multiqc_custom_methods_description = params.multiqc_methods_description ?
         file(params.multiqc_methods_description, checkIfExists: true) :
         file("$projectDir/assets/methods_description_template.yml", checkIfExists: true)
     ch_methods_description                = Channel.value(
         methodsDescriptionText(ch_multiqc_custom_methods_description))
 
-    ch_multiqc_files = ch_multiqc_files.mix(
+    ch_multiqc_extra_files = ch_multiqc_extra_files.mix(
         ch_workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))
-    ch_multiqc_files = ch_multiqc_files.mix(ch_collated_versions)
-    ch_multiqc_files = ch_multiqc_files.mix(
+    ch_multiqc_extra_files = ch_multiqc_extra_files.mix(ch_collated_versions)
+    ch_multiqc_extra_files = ch_multiqc_extra_files.mix(
         ch_methods_description.collectFile(
             name: 'methods_description_mqc.yaml',
             sort: true
         )
     )
 
-    MULTIQC (
-        ch_multiqc_files.collect(),
+    MULTIQC_GLOBAL (
+        ch_multiqc_files
+            .map { meta, file -> file }
+            .mix(ch_multiqc_extra_files)
+            .collect(),
+        ch_multiqc_config.toList(),
+        Channel.empty().toList(),
+        ch_multiqc_logo.toList()
+    )
+
+    // Generate reports by lane
+    multiqc_extra_files_per_lane = ch_multiqc_files
+        .filter { meta, sample -> meta.lane }
+        .map { meta, sample -> [lane: meta.lane] }
+        .unique()
+        .combine(ch_multiqc_extra_files)
+
+    lane_mqc_files = ch_multiqc_files
+        .filter { meta, sample -> meta.lane }
+        .mix(multiqc_extra_files_per_lane)
+        .map { meta, sample -> [ "[LANE:${meta.lane}]", meta, sample ] }
+        .groupTuple()
+        .tap { mqc_by_lane }
+        .collectFile{
+            lane, meta, samples -> [
+                "${lane}_multiqc_extra_config.yml",
+                """
+                    |output_fn_name: \"${lane}_multiqc_report.html\"
+                    |data_dir_name:  \"${lane}_multiqc_data\"
+                    |plots_dir_name: \"${lane}_multiqc_plots\"
+                """.stripMargin()
+            ]
+        }
+        .map { file -> [ (file =~ /(\[LANE:.+\])/)[0][1], file ] }
+        .join(mqc_by_lane)
+        .multiMap { lane, config, meta , samples_per_lane ->
+            samples_per_lane: samples_per_lane
+            config: config
+        }
+
+    MULTIQC_PER_LANE(
+        lane_mqc_files.samples_per_lane,
+        ch_multiqc_config.toList(),
+        lane_mqc_files.config,
+        ch_multiqc_logo.toList()
+    )
+
+    // Generate reports by group
+    multiqc_extra_files_per_group = ch_multiqc_files
```

Comment:
Correct me if I'm wrong, but when a process or workflow (in this case MULTIQC) is run on two or more queue channels, it will try to zip them and run on each pair of values.

If you don't duplicate the multiqc extra file to follow the same grouping as in the first channel, nextflow will run the lane 1 files against the first multiqc extra file, then the lane 2 files against the second multiqc extra file and so on.

This was very hard to get right, I'm all ears if you see a better way to do it :)

Reply:
I realized I confused myself with my explanation (which for sure is sign that this should be simplified). Let me try again :sweat_smile: 

When the files are provided to `MULTIQC_BY_LANE`, `lane_mqc_files.samples_per_lane` looks like this: `[list of samples for lane 1], [list of samples for lane 2], ...`. The extra multiqc files are needed for each report and need to be included in each of these lists. I thought the best solution would be to use a map over these lists and append the extra files each time, but I could never got that to work.

If you find a better way to perform this operation, I'm all for it :)